### ISSUE TITLE
[Discussion] Requiring Native Modules in the Renderer Process to be NAPI or Context Aware

### ISSUE THREAD

**Issue by MarshallOfSound** [electron/electron#18397]
### Native modules in Electron

Loading native modules in Electron has always been tricky.  It's a pain to rebuild, to link to the right `node.lib/dll`, and to ensure your `NODE_MODULE_VERSION` is correct.  And there's another problem behind the scenes that most users have never had to deal with: Node doesn't let you load multiple instances of a native module in the same process, even in different v8 contexts.  Imagine this scenario:

1. Load https://google.com in BrowserWindow
    * Process is launched.
    * Preload script runs and loads native module X
2. Load https://google.com/foo in BrowserWindow
    * Chromium decides to reuse the process
    * Preload script runs and loads native module X
    * FAILURE - native module X loaded twice in the same process

We address this in Electron by patching Chromium to create a new process for **every** navigation.  This mostly works but still has a few edge cases (e.g. #4025, #12045, #17576) and a few side effects:
* We lose all the performance gains from the Chromium plz-navigate changes :cry: 
* We have strange behavior with `window.open` and `window.opener` as we force child windows into separate processes sometimes. :cry: 
* Our patches occasionally have regressions because this isn't a process model that Chromium accepts / tests. :cry: 

We've always thought "wouldn't it be cool if we could change this". Now, with Node 12's [Worker Threads](https://nodejs.org/api/worker_threads.html), we see a way to do that.

### Worker Threads

Node ran into the same issue when they added worker threads: native modules could not be loaded in multiple workers.  Node solved this by introducing the concept of ["Context Aware"](https://nodejs.org/api/addons.html#addons_context_aware_addons) native modules.  This is how native modules tell Node that they are safe to be loaded in multiple `v8::Context`s.  [Nan](https://github.com/nodejs/nan) has a handy helper ([`NAN_MODULE_WORKER_ENABLED`](https://github.com/nodejs/nan/pull/792)) for doing this.

Context Aware modules and NAPI modules can be instantated multiple times in Node even without our patched process model. As they become more common, our patches will bring less benefit and be more redundant.

### What does this mean for Electron?

Right now, nothing. However, at some point in the future -- probably in a few major versions -- we intend to remove our process model patches. At that point, any native Node modules loaded in the renderer process **must** be either NAPI or Context Aware.

This won't happen tomorrow. It probably won't happen this year. But it will happen, so if you're using or maintaining a native module, you should start looking at the work to mark it as Context Aware. For lots of modules it's as easy as replacing `NODE_MODULE` with `NAN_MODULE_WORKER_ENABLED`. For modules that require cleanup, it can be more involved.

You may be wondering why we're changing something that (mostly) works? The main reasons are:
* **Performance**.  If we don't restart the renderer for every navigation, certain same-site navigations will be **much faster** than before. :smile:
* **Security**.  When we follow Chromium's process model, we benefit from all the the optimizations and decisions they have made around Site Isolation and related initiatives. :smile: 
* **Maintenance**.  Our patches are not easy to keep working, especially as Chromium makes more and more process model performance improvements and optimizations.  Removing the patches removes a large maintenance burden. :smile: 

### What does this look like for Native Module X?

The amount of work to become "Context Aware" will be different for every module.  For native modules that act as proxies for native getters and setters -- i.e. they just expose JS apis to acccess native APIs -- it's normally as easy as adding a `NAN_MODULE_WORKER_ENABLED` declaration.  You can find an example of that in [this pull request](https://github.com/atom/node-keytar/pull/182/files).

Some modules will need to do cleanup when the context is destroyed to ensure the module doesn't leaking memory or to cause crashes.  You can add hooks to clean up after yourself with `node::AddEnvironmentCleanupHook`  A minimal example of this can be found in [this pull request](https://github.com/CharlieHess/node-mac-notifier/pull/32/files).

### Timeline

* Land the `app.allowRendererProcessReuse` option in Electron 6
* Have the first deprecation warning land in Electron 7 for non-context aware native modules
* Have a deprecation warning land in Electron 8 for the default value of `app.allowRendererProcessReuse` to switch
* Switch the default value of `app.allowRendererProcessReuse` to `true` in Electron 9
* Deprecate the ability to change `app.allowRendererProcessReuse` in Electron 10
* Remove the ability to change `app.allowRendererProcessReuse` in Electron 14

**Comment by orange4glace** [comment]
This is cool. I just encountered such situation that needs multiple native module instances, while it is little sad that it only supports node with V8, not N-API.

**Comment by bughit** [comment]
Are native modules the only reason for the "new process for every navigation"?  The following explanation from @zcbenz seems to suggest that node can't fully tear down its one js environment per process, so a new one can take its place.

> [This is because Node is currently designed to have only one instance in one process, and it doesn't cleanup all the resources one JavaScript environment is destroyed.](6903#issuecomment-241298639)

**Comment by MarshallOfSound** [comment]
@bughit That situation has changed in recent times, as mentioned in the original issue text with the introduction of worker threads node now supports multiple instances of node per-process along with proper teardown support (this is what making your addon context aware supports).  We already teardown node environments when the `nodeIntegrationInSubFrames` flag is enabled

**Comment by bughit** [comment]
Would you consider adding an experimental option that would deactivate the "process for every navigation" mode even before you completely drop you process model patches?

Assuming that's less work then the full job.

**Comment by MarshallOfSound** [comment]
@bughit See the PR linked --> #18396 in particular the timeline outlined in that PR, what you suggest is already what's being done 😄

**Comment by AnalyzerTech** [comment]
I don't have all night for this shit. how can I opt out completely. legit code ignored and hidden. I'm probably black Listed hear.

**Comment by Savas16** [comment]
Yes

**Comment by cynx** [comment]
Can the decision to remove affinity be reviewed ? I currently don't see a way to open multiple BrowserWindows quickly enough without affinity. This would impact many developers as discussed here - https://github.com/electron/electron/issues/16319

**Comment by jacobq** [comment]
Is there a way to suppress the deprecation warning? I am upgrading to electron 7.x and plan to update my native modules soon, but right now this message floods the console messages when I run my test suite. (I tried explicitly setting `app.allowRendererProcessReuse = false;` at the beginning of the main thread, but this didn't seem to make the warning disappear.)

**Comment by bpasero** [comment]
@MarshallOfSound I second the question from https://github.com/electron/electron/issues/18397#issuecomment-558213859. Please, if you introduce a deprecation warning, also allow to turn it off. 

//cc @deepak1556

**Comment by TanninOne** [comment]
Would be brilliant if the deprecation warning also said which module it's referring to...

**Comment by TanninOne** [comment]
Thanks for the quick fix!

**Comment by Thalinda** [comment]
What should i do? i had to update to new version because of that printer problem now this

> (electron) The default value of app.allowRendererProcessReuse is deprecated, it is currently "false".  It will change to be "true" in Electron 9.  For more information please check https://github.com/electron/electron/issues/18397

**Comment by jacobq** [comment]
> What should i do? i had to update to new version because of that printer problem now this
> 
> > (electron) The default value of app.allowRendererProcessReuse is deprecated, it is currently "false".  It will change to be "true" in Electron 9.  For more information please check #18397
> 
> Now i am getting this error still it's work but i am trying to launch my app is this going to be some issue ?
> 
> > (node:11548) Electron: Loading non-context-aware native module in renderer: '\?\C:\Users\Thalinda\desktop\selectronupdated\node_modules@serialport\bindings\build\Release\bindings.node'. This is deprecated, see #18397.
> > (node:11548) Electron: Loading non-context-aware native module in renderer: '\?\C:\Users\Thalinda\desktop\selectronupdated\node_modules@serialport\bindings\build\Release\bindings.node'. This is deprecated, see #18397.

It is not an error. It is a deprecation warning. It means that in the future (e.g. maybe Electron 9, 10, ...) some things will change that will cause problems for your app unless your app is changed also. The `serialport` module will very likely be updated by its maintainers by that time, so I wouldn't worry about that one. However, it's still a good idea to take some time to read more about the changes so you don't get surprised by them later.

**Comment by Thalinda** [comment]
Ohhh, thank you soo much I have searched these issues for 6 hours now. yeah I will read it thank you very much @jacobq

**Comment by satsaiinfocom** [comment]
Fix allowRendererProcessReuse when I add this line in new version electron main.js
`app.allowRendererProcessReuse = false`;

**Comment by appurist** [comment]
Please, at least don't issue the warning if `app.allowRendererProcessReuse` is set to `true` at startup.  In that case, it's needlessly noisy and warns that the future default will be the value you're overriding it to be.  When set to true, that warning has almost no value yet can cause a lot of log pollution.

**Comment by DuBistKomisch** [comment]
`electron.app.allowRendererProcessReuse = true;` in `main.js` suppresses the warning for me with 8.0.1 @appurist maybe you need to update or are setting it at the wrong point?

**Comment by appurist** [comment]
@DuBistKomisch Thanks!  I am not sure why, what I tested before, but `electron.app.allowRendererProcessReuse = true;` avoids the message, if I place it near the top of my main process startup (which is `background.js` in my build, `main.js` is the renderer process).  We have a workaround to avoid the message in the meantime: set that value to the new default explicitly.

**Comment by tessus** [comment]
I'm getting the warning message, but it doesn't tell me which module is actually meant.

```
(node:19880) Electron: Loading non context-aware native modules in the renderer process is deprecated and will stop working at some point in the future, please see https://github.com/electron/electron/issues/18397 for more information
```

**Comment by emmkimme** [comment]
> Can the decision to remove affinity be reviewed ? I currently don't see a way to open multiple BrowserWindows quickly enough without affinity. This would impact many developers as discussed here - #16319

From my point of view (and not only mine as far as I see), rationales behind the removal of the affinity are controversial. 
-	Chromium does not manage natively such feature so we remove it
would it mean application built on Electron have to be browser designed oriented, a monolithic single window or multi-tabs, can Electron not be open to another kinds of user experience.
What about application managing independent popup windows. Removing the ability to fine-tune the way we dispatch windows over processes, will prevent Desktop applications based on multiple windows to migrate on Electron. 
We may have hundreds windows in our application, up to the client (a Trader in our case), a serious and professional person that does not want to end with hundreds of processes on his machine. A person that wants to use his multiple screens estate thanks to multiple windows.

-	Chromium already has process models which gathers windows like process-per-site
We are not browser-like which is opening pages from different sites, as most of the companies, our pages are hosted in the same domain so process-per-site does not work at all in our cases.

-	Affinity has issues (leaks, background color….)
Affinity implementation has a limited code changed and no impact on other process models. If as a developer, as a product owner,... pros are far more important than cons, I would say, it is my choice. And we are good citizen, we already contributed to Electron. If such issues became painful we will work on them. To be fair until now, it does not prevent to deliver worldwide our product to thousands of users.

In a nutshell, removing affinity will have more negative impacts than ' benefits' (if any !). It prevents Electron to be open to full Desktop-application-like that want to take advantages of a workstation in term of multiple screens and multiple windows design.

my2cents

**Comment by ArfyFR** [comment]
I updated Visual Studio Code this morning and got:
```
update#setState idle
(node:17424) Electron: Loading non context-aware native modules in the renderer process is deprecated and will stop working at some point in the future, please see https://github.com/electron/electron/issues/18397 for more information
(node:17424) Electron: Loading non context-aware native modules in the renderer process is deprecated and will stop working at some point in the future, please see https://github.com/electron/electron/issues/18397 for more information
...
```
Is this a problem ?

**Comment by indiehjaerta** [comment]
Just started a project with Electron, React and SQLite, and got this when trying it out.

`(node:18844) Electron: Loading non-context-aware native module in renderer: '\\?\...\node_modules\sqlite3\lib\binding\electron-v8.1-win32-x64\node_sqlite3.node'. This is deprecated, see https://github.com/electron/electron/issues/18397.`

How do I solve this so it won't be a problem?

**Comment by tspence** [comment]
> I'm getting the warning message, but it doesn't tell me which module is actually meant.
> 
> ```
> (node:19880) Electron: Loading non context-aware native modules in the renderer process is deprecated and will stop working at some point in the future, please see https://github.com/electron/electron/issues/18397 for more information
> ```

I would like to second this comment.  I would very much like to fix the deprecation warnings, and a tiny bit of information such as a name or a path would help a lot.

**Comment by lapagept** [comment]
Hey there!

I already have this in my main.js :
    `app.allowRendererProcessReuse = false;`

With this last line, I don't receive this warning anymore (right on!) : 
`(electron) The default value of app.allowRendererProcessReuse is deprecated, it is currently "false".  It will change to be "true" in Electron 9.  For more information please check https://github.com/electron/electron/issues/18397`

Now, how could I avoid this message in my console every time I change the window in my application?
```Electron: Loading non-context-aware native module in renderer: /Users/Flooder/Programmes/electron/console-pt-logiciel-test/node_modules/sqlite3/lib/binding/electron-v8.2-darwin-x64/node_sqlite3.node. This is deprecated, see https://github.com/electron/electron/issues/18397.```

Thanks to help!

**Comment by Pavan-Kumar122** [comment]
I am unable to use pdf2txt.py and got the same deprecated warning.

what should i do??

every time i run this command it shows this warning and open the VS code

**Comment by ap-r04** [comment]
@Pavan-Kumar122 I had a similar issue pop up an hour ago where I was trying to run idf.py for some embedded work with ESP-IDF (https://github.com/espressif/esp-idf/releases/v4.0). 

What got me moving was using:
 `python fullpath\to\idf.py ` 

The version of python I'm running:
`python -V`
`Python 3.7.3`

 Definitely not a fix tho.

**Comment by fabiospampinato** [comment]
Removing `app.allowRendererProcessReuse`, and even defaulting it to `true`, doesn't feel very safe to me for the foreseeable future. 

I've been experimenting with setting `app.allowRendererProcessReuse = true;` and I can make the app crash very easily just by using [chokidar](https://www.npmjs.com/package/chokidar), which in turns loads [fsevents](https://www.npmjs.com/package/fsevents), which [seems](https://github.com/fsevents/fsevents/blob/f2b4ba83594027a2af3b52621678c3a07bd9d4b8/src/fsevents.c#L8-L9) to be already using node's NAPI, reloading the app from the devtools, causing fsevents to be loaded again I suppose, and then triggering a filesystem change.

There are none of those "Loading non context-aware native modules (...)" errors in my console.

Is this a problem with Electron, fsevents or both?

**Comment by MarshallOfSound** [comment]
> which seems to be already using node's NAPI

Using `napi` does not make your native module multi-context safe by default, it's possible `fsevents` isn't cleaning up after itself and a file change is resulting in a freed pointer being accessed.  This will be a problem with the native module you're using, not Electron itself

**Comment by brunobg** [comment]
> Is there a way to suppress the deprecation warning?

Here's a fix that works: https://github.com/cypress-io/cypress/pull/6753

**Comment by skyperx** [comment]
root.f39bf48d.js:217 Refused to connect to 'ws://localhost:1090/' because it violates the following Content Security Policy directive: "connect-src 'self' https:".

parcelRequire.node_modules/parcel-bundler/src/builtins/hmr-runtime.js @ root.f39bf48d.js:217

**Comment by tanveerblueeast** [comment]
To run react-native code to Web I run the commands also pass reference to "whiteListedModules": ["@bluebase/core"] into webpack to convert it into web I got an error ::


Download the React DevTools for a better development experience: https://fb.me/react-devtools
vm.js:84 Uncaught /home/tanveer/BlueEast/BlueEast2/mevris-client/node_modules/react-native/Libraries/Utilities/warnOnce.js:15
const warnedKeys: {[string]: boolean} = {};
      ^^^^^^^^^^

SyntaxError: Missing initializer in const declaration
    at new Script (vm.js:84:7)
    at createScript (vm.js:258:10)
    at Object.runInThisContext (vm.js:306:10)
    at Module._compile (internal/modules/cjs/loader.js:884:26)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:986:10)
    at Module.load (internal/modules/cjs/loader.js:816:32)
    at Module._load (internal/modules/cjs/loader.js:728:14)
    at Module._load (electron/js2c/asar.js:717:26)
    at Function.Module._load (electron/js2c/asar.js:717:26)
    at Module.require (internal/modules/cjs/loader.js:853:19)

**Comment by aabuhijleh** [comment]
When not setting `app.allowRendererProcessReuse` to `false`, `fs.readdir` in the preload script's callback is never called. Is such a thing possible?

**Comment by thjxs** [comment]
Hi, I had been upgrade electron to 9, And I can't close browser window manually use `win.close()
` method, until set `app.allowRendererProcessReuse = false`

**Comment by aabuhijleh** [comment]
I think this has caused a regression in using Node.js APIs in the renderer process after page reload. #24173 #23910

**Comment by AllyTally** [comment]
Seems like requests with the `got` module completely break after reloading when this is set to true.

**Comment by implausible** [comment]
This patch from node will be necessary to run certain native node modules in Electron 11 https://github.com/nodejs/node/pull/34572. I think without this there won't be an upgrade path for modules like https://npmjs.com/package/nodegit in Electron.

**Comment by MarshallOfSound** [comment]
cc @codebytere ☝️ thoughts on a backport

**Comment by codebytere** [comment]
@implausible happy to take a look! seems reasonable

**Comment by aabuhijleh** [comment]
This change has definitely caused core Node modules to be unusable in the renderer process after a page reload (#22119) on Windows. I believe this issue will affect most if not all Electron apps when they migrate to 10. For now, there's a viable workaround (set `app.allowRendererProcessReuse=false`). But I'm not sure what will happen when Electron 11 is released with this workaround removed.

Are you aware of this issue?

Thanks!

**Comment by jacobq** [comment]
Would anyone with some expertise with worker threads be willing to lend at hand at helping make the `serialport` module compatible with this change? Progress appears to be stalled (e.g. https://github.com/serialport/node-serialport/issues/1938)

**Comment by devrim-oguz** [comment]
I also get affected on two libraries I had to use with my project with "Raspberry Pi 4 Model B" because they are hardware libraries and hard to rewrite. I added issues to their pages to support context-aware mode but it would be pretty nice if you could do something about this as well. For now, I'm running it with `app.allowRendererProcessReuse = false` option but it says you are going to remove this. If you remove it, there is no way I can run this application. Waiting for your help.

Following errors:
```

/home/alba/Interface_Automat_ArGe/node_modules/bindings/bindings.js:121 Uncaught Error: Loading non-context-aware native module in renderer: '/home/alba/Interface_Automat_ArGe/node_modules/@serialport/bindings/build/Release/bindings.node', but app.allowRendererProcessReuse is true. See https://github.com/electron/electron/issues/18397.
    at process.func [as dlopen] (electron/js2c/asar.js:140)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1196)
    at Object.func [as .node] (electron/js2c/asar.js:140)
    at Module.load (internal/modules/cjs/loader.js:981)
    at Module._load (internal/modules/cjs/loader.js:881)
    at Function.Module._load (electron/js2c/asar.js:769)
    at Module.require (internal/modules/cjs/loader.js:1023)
    at require (internal/modules/cjs/helpers.js:77)
    at bindings (/home/alba/Interface_Automat_ArGe/node_modules/bindings/bindings.js:112)
    at Object.<anonymous> (/home/alba/Interface_Automat_ArGe/node_modules/@serialport/bindings/lib/linux.js:2)
    
    
/home/alba/Interface_Automat_ArGe/node_modules/bindings/bindings.js:121 Uncaught Error: Loading non-context-aware native module in renderer: '/home/alba/Interface_Automat_ArGe/node_modules/rpio/build/Release/rpio.node', but app.allowRendererProcessReuse is true. See https://github.com/electron/electron/issues/18397.
    at process.func [as dlopen] (electron/js2c/asar.js:140)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1196)
    at Object.func [as .node] (electron/js2c/asar.js:140)
    at Module.load (internal/modules/cjs/loader.js:981)
    at Module._load (internal/modules/cjs/loader.js:881)
    at Function.Module._load (electron/js2c/asar.js:769)
    at Module.require (internal/modules/cjs/loader.js:1023)
    at require (internal/modules/cjs/helpers.js:77)
    at bindings (/home/alba/Interface_Automat_ArGe/node_modules/bindings/bindings.js:112)
    at Object.<anonymous> (/home/alba/Interface_Automat_ArGe/node_modules/rpio/lib/rpio.js:17)
```

**Comment by trusktr** [comment]
> Removing the patches removes a large maintenance burden. :smile: 

Side question: is (was) the process for maintaining those patches while across Chrome versions documented somewhere? Curious to learn how that was managed.

**Comment by nornagon** [comment]
@trusktr see here for some info on how we manage patches: https://github.com/electron/electron/blob/master/docs/development/patches.md

**Comment by DBJDBJ** [comment]
After I installed VS Code on this Windows 10 machine I can not execute js files anymore. These are javascripts admin scripts executed on Windows direct. 20 years old subsystem. But people do use them ... 

I am here because this is where Electron debug dump advised me to go

![image](https://user-images.githubusercontent.com/145155/96329982-17f18900-1052-11eb-93d8-14d41b3e4f1e.png)

**Comment by sas10x** [comment]
use electron 7.1.7 for the meantime. if you are here for serial port angular electron, using angular-electron version 6.4.1 would work.

**Comment by aminya** [comment]
How can we know which native module is causing this issue?

**Comment by vijayanto-dev** [comment]
@Marshall We are working on an internal app (Electron 5.0.1) that had affinity to tie process together. Post an upgrade to 11.0.3, since affinity is being deprecated we have moved to process-per-site, as well as allowing for renderer process to be re-used. All seem to work fine, except for an use case where we open a new-window (newGuest) to display content to the user. Crash occurs with "GUEST_WIMDOW_MANAGER_WINDOW_OPEN" denied message. Any thoughts please, would be helpful.

**Comment by implausible** [comment]
@codebytere @MarshallOfSound Hello again! So I ran into some issues with async cleanup for native node modules in Electron 11 and 12. We've patched the issue upstream in Node. It is a breaking change; however, it was not possible for anyone to compile for Electron 11 or 12 using these APIs anyway. We need to bring in this patch from Node so that modules like the ones I support are able to build for Electron 11 and up with context awareness. See this issue https://github.com/nodejs/node/issues/36349 and these 2 patches: https://github.com/nodejs/node/commit/1fe571aa0c55814e58f6efefedf69da8d358034b and https://github.com/nodejs/node/commit/1fe571aa0c55814e58f6efefedf69da8d358034b from PR https://github.com/nodejs/node/pull/37000

I hope you can see that this breaking change is probably the best approach for solving this issue (nobody can use it anyway, so what is it really breaking). Nothing short of this feat https://github.com/electron/electron/issues/27269 would provide an alternative route for integration. Right now this is seriously detrimental for NodeGit.

**Comment by MarshallOfSound** [comment]
Now that https://github.com/electron/electron/pull/26874 has landed, we can safely close this issue as "complete".  Follow up work is now possible: https://github.com/electron/electron/pull/28769 but for all intents and purposes this issue is done.

**Comment by MarshallOfSound** [comment]
Context aware modules and those modules being thread safe are two independent concepts unfortunately.  The info that doc still stands in that most native modules won't work safely in a multi threaded environment

**Comment by rathboma** [comment]
Thanks for the info. Makes sense. If that's still true, if someone (ie, Me) had to run a worker that needed native module access, how would you recommend they do it right now? Using `spawn`?

**Comment by brandonros** [comment]
What's the migration path for people who used to do `electron.app.allowRendererProcessReuse = false`?

**Comment by king2088** [comment]
electron14+ remove app.allowRendererProcessReuse, but how to loading non-context-aware native module in renderer?

**Comment by aiamuzz** [comment]
Hi,

forgive my intrusion but the problem i have been facing here ...

**Flatpak Maintainer:**
https://github.com/flathub/com.getmailspring.Mailspring/issues/3

**Mailspring Upstrema:**
https://community.getmailspring.com/t/unexplained-compose-window-popup-all-by-themselves-with-the-attachment-database-agent-js/3151/3

related to the issue discussed on this thread ?

**Comment by jacobq** [comment]
> related to the issue discussed on this thread ?

No, this is not the cause of the issue you linked. That is simply a warning regarding a change in later electron versions that would need to be accommodated in order to update electron to the latest version.

**Comment by TanninOne** [comment]
> Why is this issue closed? What are we going to do with the libraries that are not N-API compatible in the latest electron versions now?

Either you port the module, find an alternative or you only use the module from the browser process and use ipc to delegate all uses of the module to it.

**Comment by alba-ado** [comment]
By doing this electron is not pushing developers to use a safer code, they are pushing us to stuck using older versions of electrons because the projects are too big to modify and uses too many libraries.

**Comment by TanninOne** [comment]
> By doing this electron is not pushing developers to use a safer code, they are pushing us to be stuck using older versions of the electron because the projects are too big to modify and use too many libraries.

That is unfortunately my experience with the entire js/node/npm/electron environment: You need to constantly invest time on code maintenance because the entire environment is constantly changing, modules that stop being maintained commonly stop working after months or a few years, absolutely everything is built on sand.
I've worked with C++/Qt, C# and Java before, all waaaaaay more productive once a project reaches a certain size because stuff that works will continue to work and you can use your time to build new functionality instead of having to redo the same stuff all the bloody time.

If this frustrates you: Don't use electron, get out while you can. The fact that they're now doing a major release every 8 weeks and genuinely think that's a good thing tells you all you need to know.

**Comment by alba-ado** [comment]
Yeah, I fix the versions when I need to. That's why I say I'm stuck with an older electron version. However, it is nice to be able to update from time to time for improved security and bug fixes.

**Comment by TanninOne** [comment]
> I feel what you’re saying but it’s not the problem with ecosystem. Ecosystem gives you an ability to easily update dependencies. It’s up to you to deal with according problems. If you’re not having enough time/desire to resolve that, just don’t update modules, never ever.

I would be very careful with this advice, it depends on the software whether this is a good idea. Security issues become known in packages you use (directly and indirectly) and from the point they are known publicly every script kiddy under the sun will start looking for unpatched software.
Based on recent news, would you say "stick with your log4j version and never update it" is a good idea?

This is an ecosystem problem because in other ecosystems you can update the foundations to stay up-to-date with security fixes without them breaking compatibility. A library like Qt you can do updates for 10 years with no or minimal upkeep while with electron you'll be lucky to go longer than 8 weeks.
It is an ecosystem problem with npm because every package has dozens of dependencies for the most miniscule pieces of functionality (I've seen npm modules, plural, actually being used that contained a single line of code) to the point where no one is capable of monitoring the changelogs for the stuff they're using.
It's an ecosystem problem because with "professional" software if they do semantic versioning you can trust that at the very least you can install patch releases without having to expect functional changes, in the npm ecosystem all bets are off. This: "_but in practice it’s often a lie,_" is true for the npm ecosystem, not in general.

**Comment by m0059** [comment]
> ### Native modules in Electron
> Loading native modules in Electron has always been tricky. It's a pain to rebuild, to link to the right `node.lib/dll`, and to ensure your `NODE_MODULE_VERSION` is correct. And there's another problem behind the scenes that most users have never had to deal with: Node doesn't let you load multiple instances of a native module in the same process, even in different v8 contexts. Imagine this scenario:
> 
> 1. Load https://google.com in BrowserWindow
>    
>    * Process is launched.
>    * Preload script runs and loads native module X
> 2. Load https://google.com/foo in BrowserWindow
>    
>    * Chromium decides to reuse the process
>    * Preload script runs and loads native module X
>    * FAILURE - native module X loaded twice in the same process
> 
> We address this in Electron by patching Chromium to create a new process for **every** navigation. This mostly works but still has a few edge cases (e.g. #4025, #12045, #17576) and a few side effects:
> 
> * We lose all the performance gains from the Chromium plz-navigate changes 😢
> * We have strange behavior with `window.open` and `window.opener` as we force child windows into separate processes sometimes. 😢
> * Our patches occasionally have regressions because this isn't a process model that Chromium accepts / tests. 😢
> 
> We've always thought "wouldn't it be cool if we could change this". Now, with Node 12's [Worker Threads](https://nodejs.org/api/worker_threads.html), we see a way to do that.
> 
> ### Worker Threads
> Node ran into the same issue when they added worker threads: native modules could not be loaded in multiple workers. Node solved this by introducing the concept of ["Context Aware"](https://nodejs.org/api/addons.html#addons_context_aware_addons) native modules. This is how native modules tell Node that they are safe to be loaded in multiple `v8::Context`s. [Nan](https://github.com/nodejs/nan) has a handy helper ([`NAN_MODULE_WORKER_ENABLED`](https://github.com/nodejs/nan/pull/792)) for doing this.
> 
> Context Aware modules and NAPI modules can be instantated multiple times in Node even without our patched process model. As they become more common, our patches will bring less benefit and be more redundant.
> 
> ### What does this mean for Electron?
> Right now, nothing. However, at some point in the future -- probably in a few major versions -- we intend to remove our process model patches. At that point, any native Node modules loaded in the renderer process **must** be either NAPI or Context Aware.
> 
> This won't happen tomorrow. It probably won't happen this year. But it will happen, so if you're using or maintaining a native module, you should start looking at the work to mark it as Context Aware. For lots of modules it's as easy as replacing `NODE_MODULE` with `NAN_MODULE_WORKER_ENABLED`. For modules that require cleanup, it can be more involved.
> 
> You may be wondering why we're changing something that (mostly) works? The main reasons are:
> 
> * **Performance**.  If we don't restart the renderer for every navigation, certain same-site navigations will be **much faster** than before. 😄
> * **Security**.  When we follow Chromium's process model, we benefit from all the the optimizations and decisions they have made around Site Isolation and related initiatives. 😄
> * **Maintenance**.  Our patches are not easy to keep working, especially as Chromium makes more and more process model performance improvements and optimizations.  Removing the patches removes a large maintenance burden. 😄
> 
> ### What does this look like for Native Module X?
> The amount of work to become "Context Aware" will be different for every module. For native modules that act as proxies for native getters and setters -- i.e. they just expose JS apis to acccess native APIs -- it's normally as easy as adding a `NAN_MODULE_WORKER_ENABLED` declaration. You can find an example of that in [this pull request](https://github.com/atom/node-keytar/pull/182/files).
> 
> Some modules will need to do cleanup when the context is destroyed to ensure the module doesn't leaking memory or to cause crashes. You can add hooks to clean up after yourself with `node::AddEnvironmentCleanupHook` A minimal example of this can be found in [this pull request](https://github.com/CharlieHess/node-mac-notifier/pull/32/files).
> 
> ### Timeline
> * Land the `app.allowRendererProcessReuse` option in Electron 6
> * Have the first deprecation warning land in Electron 7 for non-context aware native modules
> * Have a deprecation warning land in Electron 8 for the default value of `app.allowRendererProcessReuse` to switch
> * Switch the default value of `app.allowRendererProcessReuse` to `true` in Electron 9
> * Deprecate the ability to change `app.allowRendererProcessReuse` in Electron 10
> * Remove the ability to change `app.allowRendererProcessReuse` in Electron 14

**Comment by toghrulgasimov** [comment]
same problem with electron >= 16. 
setting value of  `app.allowRendererProcessReuse` doesn't help.

**Comment by jcwii** [comment]
Considering the complexity of this thread it's seemingly lacking an ELI5 — is this currently without a workaround?

**Comment by gpetrov** [comment]
It is not about complexity is about that many open source node modules haven't switched yet to NAPI and are not Context Aware.

So now that Electron removed the `app.allowRendererProcessReuse`  switch in Electron 14, we are stuck to Electron 13 and begging the writers of the native node modules to update - which is not very effective as it is huge undertaking. See all the linked topics ....

So I really wish that Electron left the `app.allowRendererProcessReuse` in place for at least Electron 14 and 15 - this will give the authors more time as it will take at least one year to switch.

**Comment by ShruthiMallaiah** [comment]
> same problem with electron >= 16. setting value of `app.allowRendererProcessReuse` doesn't help.

Did you find any solution or workaround for this issue?

**Comment by jeevansaidasyam** [comment]
![Screenshot (28)](https://github.com/electron/electron/assets/165809442/8f07459e-a87b-454d-8ef8-4d3613265421)
![Screenshot (29)](https://github.com/electron/electron/assets/165809442/afcbb9a2-dae1-4bc3-ac50-5da3aa2ba457)
